### PR TITLE
fix query credentials of user return duplicate data

### LINKED ISSUE #647: Querying user credentials returns duplicate data
image

I add one credential but return duplicate data when querying
I will fix it.

### PR DESCRIPTION
**Pull request by baixinsui**
fixed #647

### CODE REVIEW

File: modules/credential/src/main/java/org/eclipse/xpanse/modules/credential/CredentialCenter.java [hunk 1]
```diff
@@ -146,19 +143,33 @@ public List<AbstractCredentialInfo> getCredentials(
...
         }
         return maskSensitiveValues(abstractCredentialInfos);
     }
 
+
+    private List<AbstractCredentialInfo> getAllTypeCredentials(Csp csp, String xpanseUser) {
+        List<AbstractCredentialInfo> allTypeCredentials = new ArrayList<>();
+        // Deduplication based on Csp and type and name
+        List<AbstractCredentialInfo> definedCredentialInfos =
+                pluginManager.getOrchestratorPlugin(csp).getCredentialDefinitions();
```
[the remark this answers is not in the document]
Comment (swaroopar) [reply modules/credential/src/main/java/org/eclipse/xpanse/modules/credential/CredentialCenter.java:156]: @sonatype-lift ignoreall

File: modules/credential/src/main/java/org/eclipse/xpanse/modules/credential/CredentialCenter.java [hunk 2]
```diff
@@ -146,19 +143,33 @@ public List<AbstractCredentialInfo> getCredentials(
...
+        definedCredentialInfos.stream()
+                .collect(Collectors.collectingAndThen(
+                        Collectors.toCollection(() -> new TreeSet<>(Comparator.comparing(
+                                o -> o.getCsp().name() + o.getType() + o.getName()))),
+                        HashSet::new))
+                .forEach(credentialInfo -> {
+                    AbstractCredentialInfo abstractCredentialInfo =
+                            credentialsStore.getCredential(
```
Comment (swaroopar) [inline modules/credential/src/main/java/org/eclipse/xpanse/modules/credential/CredentialCenter.java:164]: @baixinsui just to understand the fix, this method from credential store returns only one credential. From where was the duplicate exactly generated before the fix?

**Review by swaroopar — CHANGES_REQUESTED** [review]
comment added.

File: modules/credential/src/main/java/org/eclipse/xpanse/modules/credential/CredentialCenter.java (hunk 2 above)
Comment (baixinsui) [reply modules/credential/src/main/java/org/eclipse/xpanse/modules/credential/CredentialCenter.java:164]: In the plugin of csp defines a AbstractCredentialInfo with type 'variables' name 'user_passwd' like 'AK_SK' and add it into List<AbstractCredentialInfo> will happen this problem.

**Comment by swaroopar** [comment]
@baixinsui as discussed, I see two issues here. 

- [ ] we must ensure the 'name' of the credential definition is really unique for a given CSP and Credential Type. If it is not unique, we must throw an exception during bean initialization. 
- [ ] Add also 'name' to cache key. 
- [ ] Add also doc to the 'name' field to mention that it must be kept unique.

**Comment by baixinsui** [comment]
> @baixinsui as discussed, I see two issues here.
> 
> * [ ]  we must ensure the 'name' of the credential definition is really unique for a given CSP and Credential Type. If it is not unique, we must throw an exception during bean initialization.
> * [ ]  Add also 'name' to cache key.
> * [ ]  Add also doc to the 'name' field to mention that it must be kept unique.
All above I have finished.

File: modules/credential/src/main/java/org/eclipse/xpanse/modules/credential/CredentialCenter.java [hunk 3]
```diff
@@ -128,35 +122,44 @@ public List<AbstractCredentialInfo> getCredentialsByUser(String xpanseUser) {
...
+                getUserCredentials(csp, type, xpanseUser);
+        return maskSensitiveValues(userCredentialInfos);
+    }
+
+
+    private List<AbstractCredentialInfo> getUserCredentials(Csp csp, CredentialType type,
+                                                            String xpanseUser) {
+        List<AbstractCredentialInfo> userCredentials = new ArrayList<>();
+        List<AbstractCredentialInfo> definedCredentialInfos =
+                pluginManager.getOrchestratorPlugin(csp).getCredentialDefinitions();
```
[the remark this answers is not in the document]
Comment (swaroopar) [reply modules/credential/src/main/java/org/eclipse/xpanse/modules/credential/CredentialCenter.java:145]: @sonatype-lift ignore

File: modules/models/src/main/java/org/eclipse/xpanse/modules/models/credential/AbstractCredentialInfo.java [hunk 4]
```diff
@@ -87,4 +87,8 @@ public abstract class AbstractCredentialInfo {
         this.description = description;
```
Comment (swaroopar) [inline modules/models/src/main/java/org/eclipse/xpanse/modules/models/credential/AbstractCredentialInfo.java:87]: can you also please update the description of "name" field with more information?

**Review by swaroopar — CHANGES_REQUESTED** [review]
comments added.

File: modules/credential/src/main/java/org/eclipse/xpanse/modules/credential/CredentialCenter.java [hunk 5]
```diff
@@ -437,30 +429,21 @@ private List<AbstractCredentialInfo> maskSensitiveValues(
     }
```
[the remark this answers is not in the document]
Comment (baixinsui) [reply modules/credential/src/main/java/org/eclipse/xpanse/modules/credential/CredentialCenter.java:425]: Done

File: modules/models/src/main/java/org/eclipse/xpanse/modules/models/credential/AbstractCredentialInfo.java [hunk 6]
```diff
@@ -44,8 +44,9 @@ public abstract class AbstractCredentialInfo {
     @Getter
```
[the remark this answers is not in the document]
Comment (baixinsui) [reply modules/models/src/main/java/org/eclipse/xpanse/modules/models/credential/AbstractCredentialInfo.java:44]: @sonatype-lift ignoreall

File: modules/credential/src/main/java/org/eclipse/xpanse/modules/credential/CredentialCenter.java [hunk 7]
```diff
@@ -72,21 +72,26 @@ public List<CredentialType> getAvailableCredentialTypesByCsp(Csp csp) {
...
         OrchestratorPlugin plugin = pluginManager.getOrchestratorPlugin(csp);
-        List<AbstractCredentialInfo> abstractCredentialInfoList;
-        if (Objects.isNull(type)) {
-            abstractCredentialInfoList = plugin.getCredentialDefinitions();
-        } else {
-            abstractCredentialInfoList =
-                    plugin.getCredentialDefinitions().stream().filter(credential ->
-                                    Objects.equals(credential.getType(), type))
-                            .collect(Collectors.toList());
+        List<AbstractCredentialInfo> abstractCredentialInfoList = plugin.getCredentialDefinitions();
```
[the remark this answers is not in the document]
Comment (baixinsui) [reply modules/credential/src/main/java/org/eclipse/xpanse/modules/credential/CredentialCenter.java:82]: @sonatype-lift ignoreall

**Review by swaroopar — APPROVED** [review]
LGTM

**Pull request merged**